tests/python/pants_test/engine/legacy/test_graph_integration.py

```diff
@@ -0,0 +1,178 @@
+
+
+class GraphIntegrationTest(PantsRunIntegrationTest):
+
+  _SOURCES_TARGET_BASE = 'testprojects/src/python/sources'
+
+  _SOURCES_ERR_MSGS = {
+    'missing-globs': ("globs('*.a')", ['*.a']),
+    'missing-rglobs': ("rglobs('*.a')", ['**/*.a']),
+    'missing-zglobs': ("zglobs('**/*.a')", ['**/*.a']),
+    'missing-literal-files': (
+      "['nonexistent_test_file.txt', 'another_nonexistent_file.txt']", [
+      'nonexistent_test_file.txt',
+      'another_nonexistent_file.txt',
+    ]),
+    'some-missing-some-not': ("globs('*.txt', '*.rs')", ['*.rs']),
+    'overlapping-globs': ("globs('sources.txt', '*.txt')", ['*.txt']),
+  }
+
+  _BUNDLE_TARGET_BASE = 'testprojects/src/java/org/pantsbuild/testproject/bundle'
+  _BUNDLE_TARGET_NAME = 'missing-bundle-fileset'
+
+  _BUNDLE_ERR_MSGS = [
+    ("rglobs('*.aaaa', '*.bbbb')", ['**/*.aaaa', '**/*.bbbb']),
+    ("globs('*.aaaa')", ['*.aaaa']),
+    ("zglobs('**/*.abab')", ['**/*.abab']),
+    ("['file1.aaaa', 'file2.aaaa']", ['file1.aaaa', 'file2.aaaa']),
+  ]
+
+  _ERR_FMT = "WARN] In target {base}:{name} with {desc}={glob}: glob pattern '{as_zsh_glob}' did not match any files."
+
+  def _list_target_check_warnings(self, target_name):
+    target_full = '{}:{}'.format(self._SOURCES_TARGET_BASE, target_name)
+    glob_str, expected_globs = self._SOURCES_ERR_MSGS[target_name]
+
+    pants_run = self.run_pants(['list', target_full], config={
+      GLOBAL_SCOPE_CONFIG_SECTION: {
+        'glob_expansion_failure': 'warn',
+      },
+    })
+    self.assert_success(pants_run)
+
+    for as_zsh_glob in expected_globs:
+      warning_msg = self._ERR_FMT.format(
+        base=self._SOURCES_TARGET_BASE,
+        name=target_name,
+        desc='sources',
+        glob=glob_str,
+        as_zsh_glob=as_zsh_glob)
+      self.assertIn(warning_msg, pants_run.stderr_data)
+
+  @unittest.skip('Skipped to expedite landing #5769: see #5863')
+  def test_missing_sources_warnings(self):
+    for target_name in self._SOURCES_ERR_MSGS.keys():
+      self._list_target_check_warnings(target_name)
+
+  @unittest.skip('Skipped to expedite landing #5769: see #5863')
+  def test_existing_sources(self):
+    target_full = '{}:text'.format(self._SOURCES_TARGET_BASE)
+    pants_run = self.run_pants(['list', target_full], config={
+      GLOBAL_SCOPE_CONFIG_SECTION: {
+        'glob_expansion_failure': 'warn',
+      },
+    })
+    self.assert_success(pants_run)
+    self.assertNotIn("WARN]", pants_run.stderr_data)
+
+  @unittest.skip('Skipped to expedite landing #5769: see #5863')
+  def test_missing_bundles_warnings(self):
+    target_full = '{}:{}'.format(self._BUNDLE_TARGET_BASE, self._BUNDLE_TARGET_NAME)
+    pants_run = self.run_pants(['list', target_full], config={
+      GLOBAL_SCOPE_CONFIG_SECTION: {
+        'glob_expansion_failure': 'warn',
+      },
+    })
+
+    self.assert_success(pants_run)
+
+    # FIXME: this fails because we just report the string arguments passed to rglobs(), not the
```

Comment:
That feels like a separate issue than the warnings issue: would open a ticket about moving the normalization we currently do here https://github.com/pantsbuild/pants/blob/master/src/python/pants/source/wrapped_globs.py#L299 into rust, as I think that is what would be necessary to do this.

Reply:
Broke out into a comment on #5863.